File: main/java/org/goafabric/core/data/logic/PatientLogic.java
package org.goafabric.core.data.logic;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.transaction.Transactional;
import org.goafabric.core.data.controller.vo.Patient;
import org.goafabric.core.data.logic.mapper.PatientMapper;
import org.goafabric.core.data.repository.PatientRepository;

import java.util.List;

@ApplicationScoped
@Transactional
public class PatientLogic {
    private final PatientMapper patientMapper;

    private final PatientRepository patientRepository;

    public PatientLogic(PatientMapper patientMapper, PatientRepository patientRepository) {
        this.patientMapper = patientMapper;
        this.patientRepository = patientRepository;
    }

    public void deleteById(String id) {
        patientRepository.deleteById(id);
    }

    public Patient getById(String id) {
        return patientMapper.map(
                patientRepository.findById(id));
    }

    public List<Patient> findByGivenName(String givenName) {
        return patientMapper.map(
                patientRepository.findByGivenNameStartsWithIgnoreCase(givenName));
    }

    public List<Patient> findByFamilyName(String familyName) {
        return patientMapper.map(
                patientRepository.findByFamilyNameStartsWithIgnoreCase(familyName));
    }

    public Patient save(Patient patient) {
        return patientMapper.map(patientRepository.save(
                patientMapper.map(patient)));
    }

    //performance optimazation if we only nead the lastnames, otherwise stupid hibernate will fetch 1:n relations with n queries
    public List<String> searchLastNames(String search) {
        return patientRepository.findByFamilyNameStartsWithIgnoreCase(search)
                .stream().map(p -> p.familyName).toList();
    }


}

File: src/main/kotlin/org/goafabric/core/organization/logic/PatientLogic.kt
package org.goafabric.core.organization.logic

import jakarta.enterprise.context.ApplicationScoped
import jakarta.transaction.Transactional
import org.goafabric.core.organization.controller.dto.Patient
import org.goafabric.core.organization.logic.mapper.PatientMapper
import org.goafabric.core.organization.logic.phonetic.ColognePhonetic
import org.goafabric.core.organization.persistence.PatientRepository
import org.goafabric.core.organization.persistence.entity.PatientEo

@ApplicationScoped
@Transactional
class PatientLogic(
    private val mapper: PatientMapper,
    private val repository: PatientRepository
) {
    private val phonetic = ColognePhonetic()

    fun getById(id: String): Patient = mapper.map(repository.findById(id) ?: throw IllegalArgumentException("Patient not found: $id"))

    fun deleteById(id: String) = repository.deleteById(id)

    fun findByGivenName(givenName: String): List<Patient> =
        mapper.map(repository.findByGivenNameStartsWith(givenName))

    fun findByFamilyName(familyName: String): List<Patient> =
        mapper.map(repository.findByFamilyNameStartsWith(familyName))

    fun save(patient: Patient): Patient {
        val eo = mapper.map(patient)
        val withSoundex = PatientEo(
            id = eo.id,
            givenName = eo.givenName,
            givenSoundex = phonetic.encode(eo.givenName),
            familyName = eo.familyName,
            familySoundex = phonetic.encode(eo.familyName),
            gender = eo.gender,
            birthDate = eo.birthDate,
            address = eo.address,
            contactPoint = eo.contactPoint,
            version = eo.version
        )
        return mapper.map(repository.save(withSoundex))
    }

    fun findPatientNamesByFamilyName(search: String): List<Patient> =
        mapper.map(repository.findPatientNamesByFamilyNameStartsWithIgnoreCaseOrderByFamilyName(search))

    fun findByFamilyNameAndGivenName(familyName: String, givenName: String): List<Patient> =
        mapper.map(
            repository.findByFamilyNameAndGivenName(
                familyName.lowercase(),
                phonetic.encode(familyName) ?: "",
                givenName,
                phonetic.encode(givenName) ?: ""
            )
        )
}
